README
NAME
    apid - Generic REST API Daemon

SYNOPSIS
     Usage: apid [ -c file ] [ -f ] { command }
      -c file         Specify configuration file (instead of apid.conf)
      -f              Run in the foreground (don't detach)

     Possible commands are:
      start      Starts a new apid if there isn't one running already
      stop       Stops a running apid
      reload     Causes a running apid to reload it's config file.
                 Starts a new one if none is running.
      restart    Stops a running apid if one is running. Starts a new one.
      check      Check the configuration file and report the daemon state
      help       Display this usage info
      version    Display the version of apid
      debug      Starts a new apid in the foreground

DESCRIPTION
    apid is a generic http(s) daemon which can be used to provide a RESTful
    web service in front of something which isn't already web aware. If you
    already have some application server or website with a service running,
    apid is of no use for you. However, if there's some arcane, weird or
    just old computing system which is accessible by perl you want to make
    available online as a web service, then apid might be a solution.

    To use apid, you have to write a perl script which maps uris to
    handlers, so you're totally free in what you want to achieve and how to
    do it.

FEATURES
    *   supports http and https.

    *   authentication via POST vars or basic authentication

    *   decorators which you can use to enable authentication or input
        validation per uri.

    *   automatically converts incoming data (post vars, json post or query
        string) to a perl structure for easy access. Handlers return perl
        structures which will be converted automatically to json as well.

CONFIGURATION
    A config file is required for apid to work. The format is very simple,
    one option per line, the value separated by an equal sign. Empty lines
    or lines preceeded with '# are ignored.

    Possible parameters:

     host       = localhost
     port       = 4433
     map        = my.pm
     apiname    = My API
     apiversion = 0.0.1
     sslcrt     = server.crt
     sslkey     = server.key

    If sslkey or sslcrt is omitted, apid will speak http, otherwise https.
    You can configure more aspects of ssl by using IO::Socket::SSL-new()>
    parameters.

MAP SCRIPT
    The map script, in the config specified with the map parameter, controls
    the behavior of apid. In its simplest form it only contains a couple of
    handlers, here an example:

     get '/date' => sub {
       my $date = scalar localtime();
       return { date => $date };
     };

    Now, start apid:

     apid -c my.conf -f start

    And access the api function:

     % curl http://localhost:8080/date
     {"date":"Wed Oct 22 20:29:50 2014"}

    Can't be easier.

  AUTHENTICATION
    To use authentication, you have to implement a login function and you
    have to tell apid which kind of auth you want.

    Full example:

     use Authen::Simple::LDAP;

     auth basic => 'my api';

     implement login => sub {
       my($user, $pass) = @_;

       my $ldap = Authen::Simple::LDAP->new( 
         host    => 'ldap.company.com',
         basedn  => 'ou=People,dc=company,dc=net'
       );

       if ( $ldap->authenticate( $user, $pass ) ) {
         return 1; # ok
       }

       return 0; # fail
     };

     request login;
     get '/date' => sub {
       my $date = scalar localtime();
       return { date => $date };
     };

    In this case we are using basic authentication which is backed by LDAP.
    If successfull, apid will return a cookie with a session id, which can
    be used in subsequent requests. However, with basic authentication this
    is optional, you may also leave the session cookie and just put the auth
    data into every request.

   ENABLE BASIC AUTHENTICATION
     auth basic => 'my api';

    The second parameter to the auth decorator is the realm.

   ENABLE POST/REDIRECT AUTHENTICATION
     auth redirect => '/login';

    The second parameter to the auth decorator is the login uri.

    In this mode, an unauthenticated user is being redirected to the
    specified uri, where the user has to POST the username and password,
    which can either be posted as a JSON string or as query string.
    Examples:

    Post auth data as JSON string:

     curl -d "{\"user\":{\"me\":\"mypass\"}}" http://localhost:8080/login

    Post auth data directly:

     curl -d "user=me&pass=mypass" http://localhost:8080/login

    It is also possible to use a query string

     curl "http://localhost:8080/login?user=me&pass=mypass"

   LOGIN IMPLEMENTATION
    In either case, you must implement the actual login function by using
    the 'implement' decorator:

     implement login => sub { my($user, $pass) = @_; ... };

    Inside, you can use whatever you want. I'd suggest using one of the
    Authen::Simple submodules.

    The login handler must return true to indicate authentication was
    successfull.

   AUTHENTICATION DECORATOR
    To enable authentication for a specific uri, add the following decorator
    in front of it:

     request login;
     get '/date' => sub { .. };

    This has to be done for every uri handler. If you leave the decorator
    for a handler it can be accessed without authentication. Example:

     request login;
     get '/date'   => sub { .. };

     get '/uptime' => sub { .. };

     request login;
     get '/vmstat' => sub { .. };

    In this example, the uris /data and /vmstat require authentication while
    /uptime can be accessed by everyone.

  URI MAPPING
    There's only one decorator call you use to map an uri to a handler: get.
    Apid doesn't distinguish between POST, PUT, DELETE or GET requests. So,
    however the uri have been called, your handler will always be called. If
    you need to distinguish between the various request types, you have to
    do it yourself in your handler.

     get '/some/uri' => sub { my $data = shift; ... return {}; };

    The handler gets passed the submitted data as its first and only
    parameter, if present. The data is always a perl structure.

    Apid expects the handler to return a perl structure as well, which will
    be converted to JSON and returned to the client.

    There are a couple of variables which are available to each handler:

    $req
        This is a standard HTTP::Request object. In addition, if
        authentication was enabled, it contains the username of the
        authenticated client:

         $req->{user}

    $res
        This is a standard HTTP::Response object. You may modify the HTTP
        return code or add additional headers to the response as you please.

    %cfg
        This is a hash containing all options of the configuration file. It
        has been parsed by Config::General.

  INPUT VALIDATION
    Apid can validate input data automatically by using
    Data::Validate::Struct. To enable it, use the validate decorator:

     request validate => { expression => 'text' };
     get '/ps/search' => sub {
       my $data = shift;
       return &ps2a($data->{expression});
     };

    The parameter to the decorator is the validator struct required by
    Data::Validate::Struct. Please refer to the documentation there for
    details.

    If input validation fails, apid will return an error message as JSON and
    HTTP response code 403.

  AUTOMATIC DOCUMENTATION
    Usually you'll want to write the documentation for your API yourself.
    For the lazy ones, there's a documentation decorator, which you can use
    to generate it.

     request doc => 'some text';
     get '/some/uri' => sub { .. };

    If apid encounters one or more documentation decorators it generates a
    documentation which is available at /doc/.

    Beware, that this documentation is very basic, however it at least
    explains if the uri requires authentication, what kind or input it
    expects (if validation were enabled) and if authentication is required.

HELPFUL CURL COMMANDS FOR TESTING
    auth to url with login requested:

     curl -c cookies -b cookies -k -v --user USER:PASS https://localhost:4443/foo/bar

    access url when auth ok:

     curl -c cookies -b cookies -k -v https://localhost:4443/foo/bar

    post query data:

     curl -k -v -d "name=foo&year=2014" https://localhost:4443/foo/bar

    post json data:

     curl -k -v -d "{\"user\":{\"name\":\"foo\"}}" https://localhost:4443/foo/bar

    post json file 'body.json':

     curl -k -v -H "Content-Type: application/json" -d @body.json https://localhost:4443/foo/bar

    post data as query string:

     curl -k -v -d "https://localhost:4443/foo/bar?name=hans&age=2014"

    get json data:

     curl -k -v -d https://localhost:4443/foo/bar

AUTHOR
    T.v.Dein <tlinden@cpan.org>

BUGS
    Report bugs to
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-REST-Apid

SEE ALSO
    WWW::REST::Apid HTTP::Daemon HTTP::Daemon::SSL Daemon::Generic
    Config::General Data::Validate::Struct

COPYRIGHT
    Copyright (c) 2014 by T.v.Dein <tlinden@cpan.org>. All rights reserved.

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

VERSION
    apid Version 0.05.



Changelog
0.06:
	added config vars:
	 - docroot: apid does a chdir() to this directory
	 - pidfile: specify where to store the pid

0.05:
	changed namespace, made module of it,
	apid daemon only uses the module

0.01:
	initial version


Makefile.PL
use ExtUtils::MakeMaker;

my $preop =
    'perl -MPod::Markdown -e \'Pod::Markdown->new->filter(@ARGV)\' bin/apid | tee $(DISTVNAME)/README > README.md;' .
    'pod2text bin/apid | tee $(DISTVNAME)/README > README;';

WriteMakefile(
              'NAME'         => 'WWW::REST::Apid',
              'VERSION_FROM' => 'lib/WWW/REST/Apid.pm', # finds $VERSION
              'EXE_FILES'    => [ 'bin/apid' ],
	      'PREREQ_PM'    => {
				 'CGI::Cookie' => 0,
				 'Config::General' => 0,
				 'Crypt::Random' => 0,
				 'DB_File' => 0,
				 'Daemon::Generic' => 0,
				 'Data::Dumper' => 0,
				 'Data::Validate::Struct' => 0,
				 'Digest::SHA' => 0,
				 'HTTP::Daemon' => 0,
				 'HTTP::Daemon::SSL' => 0,
				 'HTTP::Request::Params' => 0,
				 'JSON' => 0,
				 'MIME::Base64' => 0,
				 'URI::Escape' => 0,
				},
              'BUILD_REQUIRES' => {
                                 'ExtUtils::MakeMaker' => 0,
                                 'LWP' => 0,
                                  },	
	      'META_MERGE' => {
			       resources => {
					     repository => 'https://github.com/TLINDEN/apid',
					    },
			      },
	      ($ExtUtils::MakeMaker::VERSION ge '6.31'? ('LICENSE'              => 'perl', ) : ()),
	      'clean'        => { FILES => 't/*.out t/test.cfg *~ */*~' },
	      MAN3PODS => {},
	      dist                => {
				      COMPRESS => 'gzip -9f',
				      SUFFIX => 'gz',
				      PREOP => $preop,
				     },
);


sample/ps.pm
# -*-cperl-*-

sub dops {
  my $filter = shift;
  my %p;

  my @lines = split /\n/, `ps axuw`;
  my @names = split /\s\s*/, shift @lines;

  foreach my $line (@lines) {
    if ($filter) {
      next if ($line !~ /$filter/);
    }
    my @vals = split /\s\s*/, $line, scalar @names;
    my %v = map { $names[$_] => $vals[$_] } 0 .. $#vals;
    $p{$v{PID}} = \%v;
  }

  return \%p;
}

sub ps2a {
  my $filter = shift;
  my $ps = &dops($filter);
  my @out = map { $ps->{$_} } sort { $a <=> $b } keys %{$ps};
  return \@out;
}

our %pw = ( tom => 123 );

auth basic => 'pslist';

implement login => sub {
  my($user, $pass) = @_;
  if (!exists $pw{$user}) {
    return 0;
  }
  else {
    if ($pass ne $pw{$user}) {
      return 0;
    }
  }
  return 1;
};

request login;
get '/ps' => sub {
  return &ps2a();
};

request login;
request validate => { pid => 'number' };
get '/ps/detail' => sub {
  my $data = shift;
  my $p = &dops();
  if (exists $p->{$data->{pid}}) {
    my $d = `ps ewww $data->{pid}`;
    my (undef, undef, undef, undef, $env) = split /\s\s*/, $d, 5;
    $out = $p->{$data->{pid}};
    foreach my $e (split /\s\s*/, $env) {
      my($var, $val) = split /=/, $e;
      $out->{env}->{$var} = $val;
    }
    return $out;
  }
  else {
    return {};
  }
};

request login;
request validate => { expression => 'text' };
get '/ps/search' => sub {
  my $data = shift;
  return &ps2a($data->{expression});
};


get '/' => sub {
  $res->header('Content-type' => 'text/html; charset=UTF-8');
  open I, "<index.html" or die "Could not open index.html: $!\n";
  my $html = join '', <I>;
  close I;
  $res->add_content($html);
  return 0;
};


1;


lib/WWW/REST/Apid.pm
#!/usr/bin/perl

package WWW::REST::Apid;

$WWW::REST::Apid::VERSION = '0.06';

use strict;
use warnings;

use Carp::Heavy;
use Carp;

use HTTP::Daemon;
use HTTP::Daemon::SSL;
use HTTP::Status qw(:constants :is status_message);
use HTTP::Request::Params;
use URI::Escape;
use Data::Dumper;
use CGI::Cookie;
use MIME::Base64;
use JSON;
use Digest::SHA qw(sha256_base64);
use Crypt::Random qw( makerandom );
use Data::Validate::Struct;
use DB_File;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS $req $res);
require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw($req $res);
@EXPORT_OK = qw($req $res);




sub new {
  my ($class, %param) = @_;
  my $type = ref( $class ) || $class;

  # default config
  my %settings = (
		  host        => 'localhost',
		  port        => 8080,
		  timeout     => 5,
		  reuseaddr   => 1,
		  sessionfile => '/tmp/apid.sessions',
		  sublogin    => sub { print "login not implemented\n"; return 0; },
		  log         => sub { return 0; },
		  authbasic   => 'WWW::REST::Apid',
		  authuri     => '',
		  foreground  => 0, # don't fork if true
		  );

  # internals
  my %intern = (
		map         => {},
		sessions    => undef, # initialized later
		server      => undef, # initialized later
	       );
  
  # override defaults
  foreach my $key (keys %param) {
    $settings{$key} = $param{$key}; 
  }

  my $self = \%settings;
  bless $self, $type;

  # interns
  foreach my $key (keys %intern) {
    $self->{$key} = $intern{$key};
  }

  $self->_init();

  return $self;
}

sub lateconfig {
  my($self, %hash) = @_;
  foreach my $key (keys %hash) {
    $self->{$key} = $hash{$key};
  }
}

sub mapuri {
  my($self, %p) = @_;

  $self->{map}->{$p{path}} = {
			      auth  => $p{doauth},
			      sub   => $p{handler},
			      valid => $p{validate} ? 
			               Data::Validate::Struct->new($p{validate}) : 0,
			     };
}



sub run {
  my $self = shift;

  $self->{log}("listening on $self->{host}:$self->{port}");

  if (! $self->{foreground}) {
    $self->{_old_sig_pipe_handler} = $SIG{'PIPE'};
    $SIG{'PIPE'} = 'IGNORE';
    $SIG{CHLD} = 'IGNORE';
  }

  while (1) {
    my $conn = $self->{server}->accept or next;

    if (! $self->{foreground}) {
      next if fork;
    }

    $self->_loadsessions();
    my ($r_port, $r_iaddr) = Socket::unpack_sockaddr_in($conn->peername);
    my $ip = Socket::inet_ntoa($r_iaddr);

    while (my $req = $conn->get_request) {
      $req->{remote_ip} = $ip;
      $req->{remote_port} = $r_port;
      $req->{path_info} = $req->uri->path;

      my $res = HTTP::Response->new;
      $res->code(200);
      $res = $self->_process($req, $res); # makes them global

      $conn->send_response($res);

      $self->{log}(join(' ', ($res->{user}, $ip, $res->code, $req->method, $req->uri->path)));
    }
    $self->_dumpsessions();
    exit if(! $self->{foreground});
  }
  $self->{log}("apid ended");
}




sub _init {
  my $self = shift;

  # check if ssl mode requested
  if (exists $self->{sslcrt} && exists $self->{sslkey}) {
    my %ssl;
    foreach my $key (keys %{$self}) {
      if ($key =~ /^SSL/) {
	$ssl{$key} = $self->{$key};
      }
    }
    $self->{server} = HTTP::Daemon::SSL->new(
					     LocalPort     => $self->{port},
					     LocalHost     => $self->{host},
					     ReuseAddr     => $self->{reuseaddr},
					     Timeout       => $self->{timeout},
					     SSL_key_file  => $self->{sslkey},
					     SSL_cert_file => $self->{sslcrt},
					     %ssl
					    ) or croak "Cannot start listener: $!\n";
  }
  else {
    $self->{server} = HTTP::Daemon->new(
					LocalPort => $self->{port},
					LocalHost => $self->{host},
					ReuseAddr => $self->{reuseaddr},
					Timeout   => $self->{timeout},	
				       ) or croak "Cannot start listener: $!\n";
  }
}








sub _authheader {
  my $self = shift;
  $res->header('WWW-Authenticate' => 'Basic realm=' . $self->{authbasic});
  $res->code(HTTP_UNAUTHORIZED);
  $res->header('Content-type' => 'application/json; charset=UTF-8');
  $res->add_content("{ \"error\": \"please authenticate\" }");
  return 0;
}

sub _doauthredir {
  my $self = shift;
  ($req, $res) = @_;
  my $data;

  if ($req->content) {
    if ($req->content =~ /^\{/) {
      eval { $data = decode_json($req->content); };
    }
    else {
      # try decoding as query
      my $query = HTTP::Request::Params->new({ req => $req });
      $data = $query->params;
      delete $data->{keywords};
    }
  }

  if ($data) {
    if (exists $data->{user} && exists $data->{pass}) {
      if ($self->{sublogin}->($data->{user}, $data->{pass})) {
	$self->_dosession($data->{user});
	$res->header('Content-type' => 'application/json; charset=UTF-8');
	$res->add_content("{ \"info\": \"authenticated\" }");
	return 1;
      }
    }
  }

  $res->code(HTTP_UNAUTHORIZED);
  $res->header('Content-type' => 'application/json; charset=UTF-8');
  $res->add_content("{ \"error\": \"please authenticate\" }");

  return 1;
}


sub _authredir {
  my $self = shift;
  $res->{target_uri} = URI::http->new($self->{authuri});
  $res->code(302);
  return 0;
}


sub _doauth {
  my $self = shift;

  if ($req->header('Cookie')) {
    my $rawcookie = $req->header('Cookie');
    if ($rawcookie =~ /^Session=(.*)$/) {
      my $session = uri_unescape($1);
      if (exists $self->{ses}->{$session}) {
	# ok, session known, user already authenticated
	my ($user, $time) = split /,/, $self->{ses}->{$session};
	if (time - $time < 86400) {
	  # ok, cookie age within bounds
	  $res->{user} = $user;
	  return 1;
	}
      }
    }
  }

  # no session
  if ($self->{authbasic}) {
    return $self->_doauthbasic();
  }
  else {
    return $self->_authredir();
  }
}

sub _doauthbasic {
  # no session, basic auth
  my $self = shift;

  my $auth = $req->header('Authorization');
  if (! $auth) {
    return $self->_authheader();
  }
  else {
    my ($basic, $b64) = split /\s\s*/, $auth;
    my $clear = decode_base64($b64);
    my($user, $pass) = split /:/, $clear;
    if (! $self->{sublogin}->($user, $pass)) {
      return $self->_authheader();
    }
    else {
      $self->_dosession($user);
      $res->header('WWW-Authenticate' => 'Basic realm="apid"');
    }
  }
  return 1;
}

sub _dosession {
  my ($self, $user) = @_;

  my $session = sha256_base64(makerandom ( Size => 512, Strength => 1 ) );
  $self->{ses}->{$session} = $user . "," . time;
  my $cookie = CGI::Cookie->new(
				-name    => 'Session',
				-expires => '+1d',
				-value   => $session);
  $res->header('Set-Cookie' => $cookie);
}

sub _process {
  my $self = shift;
  ($req, $res) = @_;

  my $fail = 1;
  my $path = '';
  my $found = 0;
  my $jsonop = JSON->new->allow_nonref;

  if (! $req->{path_info}) {
    $req->{path_info} = '/';
  }

  foreach my $path (sort { length($b) <=> length($a) } keys %{$self->{map}}) {
    if ($path eq $req->{path_info}) {
      $found = 1;
      if ($self->{map}->{$path}->{auth}) {
	if (! $self->_doauth()) {
	  last; # auth requested, user unauthenticated, else continue
	}
      }
      my $remainder = $req->{path_info};
      $remainder =~ s/\Q$path\E//;
      $req->{path_info} = $remainder;
      my $go = $self->{map}->{$path}->{sub};
      my ($put, $hash);

      if ($req->content) {
	if ($req->content =~ /^\{/) {
	  eval { $put = decode_json($req->content); };
	  if ($@) {
	    $@ =~ s/ at $0 line.*//;
	    $@ = "JSON Parser Error: $@";
	    last;
	  }
	}
	else {
	  # try decoding as query
	  my $query = HTTP::Request::Params->new({ req => $req });
	  $put = $query->params;
	  delete $put->{keywords}; 
	}
      }
      else {
	# maybe there were cgi get params
	my $query = HTTP::Request::Params->new({ req => $req });
	$put = $query->params;
	delete $put->{keywords}; 
      }

      if ($self->{map}->{$path}->{valid}) {
	my $ok;
	eval { $ok = $self->{map}->{$path}->{valid}->validate($put); };
	if (! $ok || $@) {
	  $@ =  $self->{map}->{$path}->{valid}->errstr();
	  chomp $@;
	  $@ =~ s/ at .*$//;
	  last;
	}
      }

      eval { $hash = $go->($put); };

      if (!$@) {
	if ($hash) {
	  my $json = encode_json($hash);
	  $res->add_content("$json");
	}
	$fail = 0;
      }

      last;
    }
  }

  if (!$found) {
    $res->code(404);
  }
  else {
    if (! $res->header('Content-type')) {
      $res->header('Content-type' => 'application/json; charset=UTF-8');
    }

    if ($fail) {
      $res->code(403);
      $res->add_content("{ \"error\": \"$@ $!\" }");
    }
  }

  $res->{user} = $res->{user} ? $res->{user} : '-';

  return $res;
}


sub _dumpsessions {
  my $self = shift;
  untie %{$self->{ses}};
}

sub _loadsessions {
  my $self = shift;
  tie %{$self->{ses}}, 'DB_File', $self->{sessions}, O_CREAT|O_RDWR, 0600, $DB_HASH;
}




1;



=head1 NAME

WWW::REST::Apid - Generic REST API Module

=head1 SYNOPSIS

 use WWW::REST::Apid;
 use Authen::Simple::LDAP;
 
 my $server = WWW::REST::Apid->new(
  host => 'localhost',
  port => 8080,
  apiname => 'my api',
  apiversion => '1.0',
  authbasic => 1,
  sublogin => sub { 
                     my($user, $pass) = @_;
                     my $ldap = Authen::Simple::LDAP->new( 
                     host    => 'ldap.company.com',
                     basedn  => 'ou=People,dc=company,dc=net'
                   );
                   if ( $ldap->authenticate( $user, $pass ) ) {
                     return 1; # ok
                   }
                   return 0; # fail
                 },
  log => sub { my $msg = shift; syslog('info', $msg); },
  foreground => 0,
 );
 
 $server->mapuri(path => '/', doauth => 1, handler => sub { return { msg => 'ok' } });
 
 $server->run();

=head1 DESCRIPTION

The WWW::REST::Apid module can be used to implement a REST API
for almost anything.

If you want fast and easy results, please try the L<apid> daemon,
which is shipped with the WWW::REST::Apid distribution, first.

=head1 METHODS

=head2 B<new>

The new method returns a new WWW::REST::Apid object. All parameters
are optional and will be preset with reasonable defaults.

Supported parameters:

=over

=item B<host>

The hostname or ip address where the daemon will listen to.
Default: 'localhost'.

=item B<port>

The TCP port to use. Default: 8080.

=item B<apiname>

The name of your API.

=item B<apiversion>

The version of your API.

=item B<authbasic>

Use HTTP Basic Authentication. The parameter defines the realm.

=item B<authuri>

Use HTTP POST Authentication with login uri redirect for unauthenticated
users.

=item B<sublogin>

Expects a closure as parameter. The closure gets two parameters supplied
during the login process: the username and the password supplied by the
client in clear text.

If authentication shall succeed, return 1, otherwise 0.

Default: returns always false.

=item B<log>

Logging function to use, expects a closure as parameter. One parameter 
will be given to the closure: the log message. Put it where ever you
want.

Default: ignore.

=item B<foreground>

If set to true, the daemon doesn't fork a child process for new
incoming connections, which it does otherwise. If you work with
a preforking system as L<Any::Daemon::HTTP>, then set it to true. If
you use something like L<Generic::Daemon>, set it to false.

Default: false.

=item B<sslcrt> AND B<sslkey>

If both are given, files are expected. B<sslcrt> must be a X509 PEM
encoded SSL certificate, B<sslkey> must be a PEM encoded SSL unencrypted
private key for the certificate.

=item B<IO::Socket::SSL->new() parameters>

Any parameter starting with 'SSL' will be fed unaltered to IO::Socket::SSL->new().

=back

=head2 B<lateconfig>

Supply any of the above mentioned parameters at some later point,
which allows to re-configure certain aspects of the daemon. Some variables
cannot be changed once the daemon runs, especially the host and port
variables.

Expects the parameters as a hash. Example:

 $server->lateconfig(authuri => '/login');

=head2 B<mapuri>

The B<mapuru> method is the heart of the module. It expects hash parameters.

Example:

 $server->mapuri(path => '/', doauth => 1, handler => sub { return { msg => 'ok' } });

The following parameters are supported:

=over

=item B<path>

Required: the uri path which shall be mapped to some action.

=item B<handler>

Required: closure is expected as parameter. The closure gets as its only
argument a hash reference supplied which contains data posted by
a client (either via POST, PUT or GET query params).

It is expected to return a hash reference with results again.

JSON conversion will be done automatically.

You can access the current L<HTTP::Request> object within the
handler by using the variable B<$req> and the L<HTTP::Response>
object with B<$res>.

=item B<doauth>

Optional: turn on authentication for this particular path.

The B<sublogin> closure must be imlemented.

=item B<valid>

Optional: a hash reference describing the input validation
using the notation of L<Data::Validate::Struct>. If defined,
data posted by clients will be validated and if found to be
invalid, an error will be returned.

=back

=head2 B<run>

Finally, start the server.

This method never returns.

=head1 AUTHOR

T.v.Dein <tlinden@cpan.org>

=head1 BUGS

Report bugs to
http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-REST-Apid

=head1 SEE ALSO

L<apid>
L<HTTP::Daemon>
L<HTTP::Daemon::SSL>
L<Daemon::Generic>
L<Config::General>
L<Data::Validate::Struct>

=head1 COPYRIGHT

Copyright (c) 2014 by T.v.Dein <tlinden@cpan.org>.
All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 VERSION

apid Version 0.06.

=cut

